vpr/src/place/place.cpp

```diff
@@ -1775,45 +2088,55 @@ static void comp_td_connection_delays(const PlaceDelayModel* delay_model) {
     }
 }
 
-/* Update the connection_timing_cost values from the temporary *
- * values for all connections that have changed.                   */
-static void commit_td_cost(const t_pl_blocks_to_be_moved& blocks_affected) {
-    auto& cluster_ctx = g_vpr_ctx.clustering();
+//Commit all the setup slack values from the PlacerSetupSlacks class.
+//This incremental routine will be correct if and only if it is called
+//immediately after each time update_setup_slacks_and_criticalities
+//updates the setup slacks (i.e. do_update_setup_slacks = true)
+static void commit_setup_slacks(const PlacerSetupSlacks* setup_slacks) {
+    const auto& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
 
-    /* Go through all the blocks moved. */
-    for (int iblk = 0; iblk < blocks_affected.num_moved_blocks; iblk++) {
-        ClusterBlockId bnum = blocks_affected.moved_blocks[iblk].block_num;
-        for (ClusterPinId pin_id : cluster_ctx.clb_nlist.block_pins(bnum)) {
-            ClusterNetId net_id = cluster_ctx.clb_nlist.pin_net(pin_id);
+    //Incremental: only go through sink pins with modified setup slack
+    auto clb_pins_modified = setup_slacks->pins_with_modified_setup_slack();
+    for (ClusterPinId pin_id : clb_pins_modified) {
+        ClusterNetId net_id = clb_nlist.pin_net(pin_id);
+        size_t pin_index_in_net = clb_nlist.pin_net_index(pin_id);
 
-            if (cluster_ctx.clb_nlist.net_is_ignored(net_id))
-                continue;
-
-            if (cluster_ctx.clb_nlist.pin_type(pin_id) == PinType::DRIVER) {
-                //This net is being driven by a moved block, recompute
-                //all point to point connections on this net.
-                for (size_t ipin = 1; ipin < cluster_ctx.clb_nlist.net_pins(net_id).size(); ipin++) {
-                    connection_delay[net_id][ipin] = proposed_connection_delay[net_id][ipin];
-                    proposed_connection_delay[net_id][ipin] = INVALID_DELAY;
-                    connection_timing_cost[net_id][ipin] = proposed_connection_timing_cost[net_id][ipin];
-                    proposed_connection_timing_cost[net_id][ipin] = INVALID_DELAY;
-                }
-            } else {
-                //This pin is a net sink on a moved block
-                VTR_ASSERT_SAFE(cluster_ctx.clb_nlist.pin_type(pin_id) == PinType::SINK);
+        connection_setup_slack[net_id][pin_index_in_net] = setup_slacks->setup_slack(net_id, pin_index_in_net);
+    }
+}
 
```

Comment:
Explain what this routine is doing.  Confirming the cached setup slacks match what you get if you ask the timing analyzer?  Also explain when it is valid to expect these to match (timing_update with any new delays hasn't happened since the cached setup slacks were created?)